jvm/src/common/Language/Java/Internal.hs

```diff
@@ -155,3 +160,20 @@ getStaticFieldAsJValue retsing cname fname = do
     SPrim "double" -> JDouble <$> getStaticDoubleField klass field
     SVoid -> fail "getStaticField cannot yield an object of type void"
     _ -> JObject <$> getStaticObjectField klass field
+
+mkVariadic :: TH.TypeQ -> (TH.TypeQ -> TH.TypeQ -> [TH.PatQ] -> TH.ExpQ -> TH.ExpQ -> TH.DecsQ) -> TH.DecsQ
+mkVariadic retty k = fmap concat $ for [0..32 :: Int] $ \n -> do
+    let -- Coercible type class and Ty associated type defined in downstream module.
+        coercible = TH.conT (TH.mkName "Coercible")
```

Comment:
This approach is quite fragile -- especially given that there may reasonably be the wrong `Coercible` in scope. It is possible to concretely specify the defining module, etc., for a name, even before it's in scope. See the `mk_name_...` functions in https://github.com/goldfirere/singletons/blob/master/src/Data/Singletons/Names.hs

Reply:
As noted above, `mkVariadic` should only be used in `Language.Java.Unsafe`, not anywhere else (and certainly not by the user). That said, if there is something more robust than `mkName` (which is only necessary because of stage restriction collateral damage), might as well. I noticed `mkNameG_tc` and friends in the template-haskell API docs, but they're not documented. What does it do, and how does it differ from `mkName`?

Reply:
Those functions make names tagged with a package and a module; `mkName` makes a name that unhygienically refers to whatever is in scope. I don't think `mkNameG_tc` was really meant to be exported from TH, but it works well in this scenario.